### ISSUE TITLE
aws-vault does not reuse existing session (regression in v5.1)

### ISSUE THREAD

**Issue by vaskevich** [99designs/aws-vault#525]
I have a pretty standard setup in AWS: a read-only IAM user along with an administrative role that can be assumed with an MFA device attached to the read-only user. Things were working fine until the later versions of aws-vault (repros on v5.3.1) - where the session doesn't seem to be reused properly, and as a result I get prompted for a new MFA code each time.

Here's my `~/.aws/config`:
```
[profile read_only]
region = us-west-2

[profile admin]
source_profile = read_only
role_arn = arn:aws:iam::1234567890123:role/OrganizationAccountAccessRole
mfa_serial = arn:aws:iam::1234567890123:mfa/ReadOnlyUser
```

The following happens on v5.1.2:
```
➜  aws-vault --debug exec admin -- aws sts get-caller-identity
2020/02/16 16:48:17 [keyring] Considering backends: [keychain pass file]
2020/02/16 16:48:17 Loading config file /Users/oleg/.aws/config
2020/02/16 16:48:17 Parsing config file /Users/oleg/.aws/config
2020/02/16 16:48:17 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:48:17 [keyring] Found 3 results
2020/02/16 16:48:17 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:48:17 [keyring] Found 3 results
2020/02/16 16:48:17 profile read_only: using stored credentials 
2020/02/16 16:48:17 profile read_only: not using GetSessionToken because MFA serial doesn't match with profile 'admin'
2020/02/16 16:48:17 profile admin: using AssumeRole (using MFA)
Enter token for arn:aws:iam::1234567890123:mfa/ReadOnlyUser: 123123
2020/02/16 16:48:22 Looking up keyring for 'read_only'
2020/02/16 16:48:22 [keyring] Querying keychain for service="aws-vault", account="read_only", keychain="aws-vault.keychain"
2020/02/16 16:48:22 [keyring] Found item "aws-vault (read_only)"
2020/02/16 16:48:22 Generated credentials ****************ABCD using AssumeRole, expires in 59m59.26429s
2020/02/16 16:48:22 Setting subprocess env: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
2020/02/16 16:48:22 Setting subprocess env: AWS_SESSION_TOKEN, AWS_SECURITY_TOKEN
2020/02/16 16:48:22 Setting subprocess env: AWS_SESSION_EXPIRATION
```

Rerunning the second time:
```
➜  aws-vault --debug exec admin -- aws sts get-caller-identity
2020/02/16 16:48:24 [keyring] Considering backends: [keychain pass file]
2020/02/16 16:48:24 Loading config file /Users/oleg/.aws/config
2020/02/16 16:48:24 Parsing config file /Users/oleg/.aws/config
2020/02/16 16:48:24 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:48:24 [keyring] Found 3 results
2020/02/16 16:48:24 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:48:24 [keyring] Found 3 results
2020/02/16 16:48:24 profile read_only: using stored credentials 
2020/02/16 16:48:24 profile read_only: not using GetSessionToken because MFA serial doesn't match with profile 'admin'
2020/02/16 16:48:24 profile admin: using AssumeRole (using MFA)
Enter token for arn:aws:iam::1234567890123:mfa/ReadOnlyUser: XXXXXX <C-c>
```

On the other hand, with version `v5.0.1`:
```
➜  aws-vault --debug exec admin -- aws sts get-caller-identity
2020/02/16 16:54:14 [keyring] Considering backends: [keychain pass file]
2020/02/16 16:54:14 Loading config file /Users/oleg/.aws/config
2020/02/16 16:54:14 Parsing config file /Users/oleg/.aws/config
2020/02/16 16:54:14 Using GetSessionToken + AssumeRole for credentials
2020/02/16 16:54:14 Looking for sessions for read_only
2020/02/16 16:54:14 Looking up all keys in keyring
2020/02/16 16:54:14 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:54:14 [keyring] Found 3 results
Enter token for arn:aws:iam::1234567890123:mfa/ReadOnlyUser: 123123
2020/02/16 16:54:19 Looking up keyring for read_only
2020/02/16 16:54:19 [keyring] Querying keychain for service="aws-vault", account="read_only", keychain="aws-vault.keychain"
2020/02/16 16:54:21 [keyring] Found item "aws-vault (read_only)"
2020/02/16 16:54:22 Generated credentials ****************ABCD using GetSessionToken, expires in 7h59m59.607383s
2020/02/16 16:54:22 Writing session for read_only to keyring: "session,XXXXXXXXXXXXXX,YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY,11111111111"
2020/02/16 16:54:22 [keyring] Checking keychain status
2020/02/16 16:54:22 [keyring] Keychain status returned nil, keychain exists
2020/02/16 16:54:22 [keyring] Keychain item trusts keyring
2020/02/16 16:54:22 [keyring] Adding service="aws-vault", label="aws-vault session for read_only", account="session,XXXXXXXXXXXXXX,YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY,11111111111", trusted=true to osx keychain "aws-vault.keychain"
2020/02/16 16:54:22 Generated credentials ****************ABCD using AssumeRole, expires in 59m59.463835s
2020/02/16 16:54:22 Setting subprocess env: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
2020/02/16 16:54:22 Setting subprocess env: AWS_SESSION_TOKEN, AWS_SECURITY_TOKEN
➜  Downloads aws-vault --debug exec admin -- aws sts get-caller-identity
2020/02/16 16:54:28 [keyring] Considering backends: [keychain pass file]
2020/02/16 16:54:28 Loading config file /Users/oleg/.aws/config
2020/02/16 16:54:28 Parsing config file /Users/oleg/.aws/config
2020/02/16 16:54:28 Using GetSessionToken + AssumeRole for credentials
2020/02/16 16:54:28 Looking for sessions for read_only
2020/02/16 16:54:28 Looking up all keys in keyring
2020/02/16 16:54:28 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/02/16 16:54:28 [keyring] Found 4 results
2020/02/16 16:54:28 Session "session,XXXXXXXXXXXXXX,YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY,11111111111" expires in 7h59m53.909137s
2020/02/16 16:54:28 [keyring] Querying keychain for service="aws-vault", account="session,XXXXXXXXXXXXXX,YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY,11111111111", keychain="aws-vault.keychain"
2020/02/16 16:54:28 [keyring] Found item "aws-vault session for read_only"
2020/02/16 16:54:28 Re-using cached credentials ****************ABCD generated from GetSessionToken, expires in 7h59m53.883124s
2020/02/16 16:54:28 Generated credentials ****************ABCD using AssumeRole, expires in 59m59.368265s
2020/02/16 16:54:28 Setting subprocess env: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
2020/02/16 16:54:28 Setting subprocess env: AWS_SESSION_TOKEN, AWS_SECURITY_TOKEN
```

Is there anything I'm doing wrong here? In the meantime the workaround for me is to just use this older version.

**Comment by jordandcarter** [comment]
I confirm the same behaviour. 5.3.1 and 5.2.0 don't create the session whereas 5.0.1 does when using file backend on Ubuntu Server 18.04.

**Comment by mtibben** [comment]
There is a workaround - put `mfa_serial = ...` in `[read_only]`.  But yes we need to work out a better way to infer this

**Comment by mtibben** [comment]
Or if you want to avoid MFA prompts for the read_only user
```
[default]
region = us-west-2

[profile user]
mfa_serial = arn:aws:iam::1234567890123:mfa/ReadOnlyUser

[profile read_only]
source_profile = user

[profile admin]
source_profile = user
role_arn = arn:aws:iam::1234567890123:role/OrganizationAccountAccessRole
mfa_serial = arn:aws:iam::1234567890123:mfa/ReadOnlyUser
```

**Comment by jfirebaugh** [comment]
> There is a workaround - put mfa_serial = ... in [read_only]

This doesn't work for me. I have:

```
[profile default]
region = us-west-2
mfa_serial = <serial>

[profile admin]
source_profile = default
mfa_serial = <same serial>
role_arn = <admin arn>
```

I'm still prompted for a token on every `aws-vault exec admin`.

**Comment by pizza0rodeo** [comment]
> This doesn't work for me. I have:
> 
> ```
> [profile default]
> region = us-west-2
> mfa_serial = <serial>
> ```
Assuming you are talking about ~/.aws/config, you should change:
```
[profile default]
```
...to:
```
[default]
```
Or, if you're talking about ~/.aws/credentials, I found that aws-vault doesn't seem to ingest that file the same way it does ~/.aws/config, so I had to move my configuration to ~/.aws/config.

**Comment by dj5k** [comment]
> There is a workaround - put `mfa_serial = ...` in `[read_only]`. But yes we need to work out a better way to infer this

This is what worked for me.

@vaskevich if you look closely, your output says:

> 2020/02/16 16:48:17 profile read_only: not using GetSessionToken because MFA serial doesn't match with profile 'admin'

I had the same issue because I removed the mfa_serial from my default profile when upgrading aws-vault from 4.7.1 to 5.3.2. But for now it seems I need to keep it.

**Comment by jonscheiding** [comment]
I think I am experiencing the same issue though with slightly different symptoms.

Say my config looks like this:

```
[profile master]
mfa_serial = arn:aws:iam::1234567890123:mfa/jonscheiding

[profile dev]
source_profile = master
role_arn = arn:aws:iam::1234567890123:role/DevelopmentRole
```

I do `aws-vault add master`, then `aws-vault list`.  

In 4.7.1 I get:

```
Profile                  Credentials              Sessions                 
=======                  ===========              ========                 
dev                      master                   -                        
master                   master                   -                        
```

In 5.3.2 I get:

```
Profile                  Credentials              Sessions                 
=======                  ===========              ========                 
dev                      -                        -                        
master                   master                   -                        
```

In 5.3.2 if I try to run `aws-vault exec dev --json`, I get an error:

```
aws-vault: error: exec: Failed to get credentials for dev: AccessDenied: User: arn:aws:iam::1234567890123:user/jonscheiding is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::1234567890123:role/DevelopmentRole
	status code: 403, request id: 00000000-0000-0000-0000-000000000000
```

Which I guess is due to our security configuration rejecting me because I didn't provide an MFA code.

 Adding the `mfa_serial` line to my `dev` profile in the config file seems to fix the problem (though the output from `aws-vault list` is still different).  I don't love this but it's livable.

**Comment by mtibben** [comment]
> Adding the mfa_serial line to my dev profile in the config file seems to fix the problem (though the output from aws-vault list is still different). I don't love this but it's livable.

This is the expected behaviour and same behaviour as the aws-cli

**Comment by jonscheiding** [comment]
> This is the expected behaviour and same behaviour as the aws-cli

Oh 😞 

I thought that the CLI would "inherit" the `mfa_serial` setting from the source profile, but just tested it and it doesn't seem to.  It would be nice if aws-vault did this (and it seems it used to) but if the CLI itself doesn't then this makes sense.

**Comment by mtibben** [comment]
Yep this is a common gotcha. I've made a note of this in https://github.com/99designs/aws-vault/blob/master/USAGE.md#gotchas-with-mfa-config

**Comment by mtibben** [comment]
This should be fixed by #569
